src/core/mempool.rs

```diff
@@ -476,6 +574,9 @@ const MEMPOOL_INDEXES: &'static [&'static str] = &[
 
 pub struct MemPoolDB {
     pub db: DBConn,
+    /// A pool (of size 1) of Mutex guarded database connections. This is a hack until we decide
+    /// the most "rust"-consistent way of passing a db connection to each mempool Iterator.
+    db_connection_pool: Arc<Mutex<DBConn>>,
```

Comment:
Note: This is a discussion point, I'm sure!

Reply:
Having looked at this code, it appears that this is only needed in order to support instantiating a `TransactionPageCursor`.  You should not need to do this.  Instead, you should do one of the following:

* Create a separate, owned database connection for the iterator.

* Borrow the `db` member of `MemPoolDB` and give `TransactionPageCursor` a lifetime parameter to ensure that it can't outlive the `MemPoolDB` it was built from.

* Implement `Iterator` for `MemPoolDB` so you don't have to borrow or copy state out of it.

* Convert `db` into `Optional<DBConn>`, take the `db` member from `MemPoolDB` to instantiate the `TransactionPageCursor`, and then recover it somehow when it goes out of scope.

* Do not use the Rust iterator trait, but instead do something like `with_db<..>(..)` I described earlier to load the next "page" of transactions.

I think the first option is probably the easiest.

Reply:
I was thinking to create an owned connection object, and pass it to the iterator.

But, iterators are being created and destroyed constantly. The miner more or less starts a new iteration once one is done.

The pattern I was going with is the "connection pool", which is pretty common:
* https://www.cockroachlabs.com/blog/what-is-connection-pooling/

Reply:
I uploaded another alternative...

I rewrote the `TestPeer` so that `MemPoolDB` doesn't get sent across threads, and so doesn't require `Send` or `Sync` to be implemented. That test framework was the only thing requiring this, so prod doesn't need it, so let's just unwind this restriction.

Thus, we can now keep a reference to `Rc<DBConn>`, instead of `Arc<Mutex<DBConn>>`. The removal of the mutex means nothing is checked at runtime, everything is checked at compile time, and there is no new risk of deadlock.

I kept then two copies of the db (a simple kind of pool), one of which can be borrowed as mutable, and the other is a read-only `Rc`.

Reply:
You should not need to use any sort of reference counting or smart pointer to store a `DBConn`.  You should be able to either borrow the mempool's connection in the iterator implementation, or create an owned connection inside the iterator.

If you find yourself trying to use `Arc`, `Rc`, `RefCell`, etc. to get around the borrow checker, then it's a sign that your code needs to be refactored.  The Rust compiler makes a lot of very useful static guarantees about the code's correctness with respect to resource acquisition and release (be it memory, file descriptors, etc.) if you can express them in terms of lifetimes.  Using reference-counting precludes asserting these static guarantees end-to-end, so they should be avoided unless it's absolutely necessary.

Reply:
ok.. `Rc` is gone!... even though I miss it :(